bundles/org.eclipse.osgi/felix/src/org/apache/felix/resolver/Candidates.java

```diff
@@ -912,7 +912,7 @@ public ResolutionError prepare()
                             {
                                 // If the original capability is from the host, then
                                 // we just need to replace it in the shadow list.
-                                getShadowList(r).replace(origCap, c);
+                                m_candidateMap.get(r).replace(origCap, c);
```

Comment:
This will be problematic if a fragment capability is attached to this host which is also a candidate for the given requirement `r`.  The reason is because it will become possible for the `ResolveContext` to get handed the modified `List<Capability>` with the `insertHostedCapability` which now will have some entries that may be wrapped and some that are not.  Yet the javadoc for `insertHostedCapability` says the list must be 
> The list returned from  {@link #findProviders(Requirement)}.

So basically we need to make sure the `ResolveContext` only gets a list of capabilities that it returned from `findProviders`, we cannot insert our internal wrappers and expose the `ResolveContext` to these internal types that it does not know about.

Reply:
So you mean the list must be the original list returned from that method correct? Is there any test or reasoning for this? If I look at the implementation, the list is defiantly modified by the ShaddowList and copied already but I'll double check that.

Reply:
Or must not the `List` be the same but only the items must only be the ones from `findProviders` correct?

Reply:
> Or must not the `List` be the same but only the items must only be the ones from `findProviders` correct?

That is correct, the objects in the list itself must have come from `findProviders`.

Reply:
If I recall, the order of the items should also be the same because they are supposed to be in preference order of what the resolver will try first to satisfy the requirement and that original order of the capabilities is supposed to be controlled by the order `findProviders` returned.